internal/mesh/internal/controllers/sidecarproxy/builder/service_port_info.go

```diff
@@ -0,0 +1,110 @@
+package builder
+
+import (
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+)
+
+// servicePortInfo is a struct used by the destination builder so that it can pre-process
+// what is the service mesh port and what are the distinct ports across the endpoints.
+// This pre-processing reduces the iterations during the BuildDestinations processing
+// would need to get the ports, check if one was the mesh port, know whether it already
+// has a cluster created for it and so on.
+
+type servicePortInfo struct {
+	// meshPortName is the name of the port with Mesh Protocol.
+	meshPortName string
+	// meshPort is the port with Mesh Protocol.
+	meshPort *pbcatalog.WorkloadPort
+	// servicePorts are the distinct ports that need clusters and routers for given destination.
+	// servicePorts do not include the mesh port and are called servicePorts because they
+	// belong to the service.
+	servicePorts map[string]*pbcatalog.WorkloadPort
+}
+
+// newServicePortInfo builds a servicePointInfo given a serviceEndpoints struct.
```

Comment:
Can we capture the union logic across addresses and why we need to do that in this comment? That way when you hover over this function call in an IDE you can see why we need to dedup the ports and ensure only the common/union ports are used.

From what I've understood, (please correct me if I'm wrong) if you're reaching api:8080, then every workload referenced by service api needs to support port 8080, otherwise we'd be erroring on some traffic. So when an destination is referenced, we can only add clusters/matches/etc for ports that are common across all the workloads/addresses for a service.

Reply:
Great call.  Yes, what you are saying is what I was trying to capture.  If this is incorrect or there is more nuance, I'm happy to change it, but this is what I was trying to capture.  

I've added a comment in ac693a9c42f5a4f8281d0a7200323fa96ccc5134 for `newService`